pkg/registrymirror/registrymirrorsstore.go

```diff
@@ -0,0 +1,325 @@
+	_, err = f.Write(newData.Bytes())
+	if err != nil {
+		return errors.Wrap(err, "could not write bytes to file")
+	}
+
+	// Close file before defer to ensure close errors are surfaced.
+	if err := f.Close(); err != nil {
+		return errors.Wrap(err, "could not close file")
+	}
+
+	// Invalidate the registries cache so that the updated file will be read on next invocation of PullSources.
+	sysregistriesv2.InvalidateCache()
+
+	log.Debugf("Successfully updated the registries config at %q with %v icspRules, %v idmsRules, %v itmsRules",
+		s.configPath, len(icspRules), len(idmsRules), len(itmsRules))
+
+	return nil
+}
+
+// PullSources will return image references in the order they should be attempted when pulling the image.
+// The returned slice may (or may not) include the source image based on the registries config.
+// Source image must be fully qualified (include the registry hostname).
+func (s *FileStore) PullSources(srcImage string) ([]string, error) {
+	// FindRegistry will parse the registries config file and cache it for future usage.
+	reg, err := sysregistriesv2.FindRegistry(s.systemContext, srcImage)
+	// Using errors.Is instead of os.IsNotExist because the return from FindRegistry
+	// wraps the error in a type that os.IsNotExist ignores.
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return nil, errors.Wrap(err, "failed finding registry")
+	}
+
+	if reg == nil {
+		return []string{srcImage}, nil
+	}
+
+	// ParseNamed assumes srcImage is a fully qualified references (contains a hostname), will produce error
+	// otherwise. In the future if support for short names is needed, use reference.ParseNormalizedNamed instead.
+	ref, err := reference.ParseNamed(srcImage)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create image reference")
+	}
+
+	pullSrcs, err := reg.PullSourcesFromReference(ref)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not pull sources from reference")
+	}
+
+	srcs := make([]string, 0, len(pullSrcs))
+	for _, src := range pullSrcs {
+		ref := src.Reference.String()
+
+		if ref == srcImage && reg.Blocked {
+			// The registries config states the src registry should not be contacted.
+			continue
+		}
+
+		srcs = append(srcs, ref)
+	}
+
+	return srcs, nil
+}
+
+// UpsertImageContentSourcePolicy will store a new/updated ImageContentSourcePolicy.
+func (s *FileStore) UpsertImageContentSourcePolicy(icsp *operatorV1Alpha1.ImageContentSourcePolicy) error {
+	s.upsertImageContentSourcePolicyWithLock(icsp)
+	return s.updateConfig()
+}
+func (s *FileStore) upsertImageContentSourcePolicyWithLock(icsp *operatorV1Alpha1.ImageContentSourcePolicy) {
+	s.ruleRWMutex.Lock()
+	defer s.ruleRWMutex.Unlock()
+
+	s.icspRules[icsp.GetUID()] = icsp
+}
+
+// DeleteImageContentSourcePolicy will delete an ImageContentSourcePolicy from the store if it exists.
+func (s *FileStore) DeleteImageContentSourcePolicy(uid types.UID) error {
+	s.deleteImageContentSourcePolicyWithLock(uid)
+	return s.updateConfig()
```

Comment:
should we consider skipping this if nothing was actually deleted? same comment for the other deletes

Reply:
We could, in my testing a delete causes the nodes to drain and sensor to restart so it shouldn't cause issues as is.

Reply:
I made the change to check if delete does anything, and added a test for the same.